### ISSUE TITLE
memory from large list of data.table can not be reclaimed by gc()

### ISSUE THREAD

**Issue by mb706** [Rdatatable/data.table#6301]
When I create a large number of small `data.table` objects, these take up much more space than equivalent `data.frame`s. The memory does not appear to get freed again after the corresponding variables are removed, even after multiple `gc()` calls.

To reproduce, run the following in an `R --vanilla` session (runs for about 2 minutes):

```r
library("data.table")
f <- function() { replicate(1e5, as.data.table(as.list(sapply(letters, function(x) TRUE))), simplify = FALSE) ; NULL }
f()
#> NULL
gc() ; gc() ; gc() ; gc()
#>             used   (Mb) gc trigger   (Mb)  max used   (Mb)
#> Ncells    828909   44.3    4659592  248.9   5824490  311.1
#> Vcells 211064732 1610.3  351555947 2682.2 214480283 1636.4
#>          used (Mb) gc trigger   (Mb)  max used   (Mb)
#> Ncells 328916 17.6    3727674  199.1   5824490  311.1
#> Vcells 664761  5.1  281244758 2145.8 214480283 1636.4
#>          used (Mb) gc trigger   (Mb)  max used   (Mb)
#> Ncells 328916 17.6    2982140  159.3   5824490  311.1
#> Vcells 664761  5.1  224995807 1716.6 214480283 1636.4
#>          used (Mb) gc trigger   (Mb)  max used   (Mb)
#> Ncells 328916 17.6    2385712  127.5   5824490  311.1
#> Vcells 664761  5.1  179996646 1373.3 214480283 1636.4
```

and check the "resident set size" / memory usage using tools like `htop`. For me, the R process takes up about 2GB, even though the garbage collector was called and there should be no more references to the created `data.table` objects. Notice how the `gc()` memory report is not aware that R is using this amount of memory.

When I run this with `data.frame()` instead of `data.table()`, the  process only grows to about 160MB. When using `1e6` instead of `1e5` for `data.frame`, the process grows to about 1GB. So `data.table()` seems to be using about 10KB (2 GB for 1e5 tables) per table, whereas `data.frame()` uses only 1KB (1GB for 1e6). 

It does not really appear to be a memory leak: when I run `f()` again, it does not end up using twice the amount of memory. I am still wondering what is going on here, and if there is a way to reclaim this memory.

Running on Linux, maybe this is different on other systems.

### Output of `sessionInfo()`

```
> sessionInfo()
R version 4.3.2 (2023-10-31)
Platform: x86_64-redhat-linux-gnu (64-bit)
Running under: Fedora Linux 38 (Thirty Eight)

Matrix products: default
BLAS/LAPACK: FlexiBLAS OPENBLAS-OPENMP;  LAPACK version 3.11.0

locale:
 [1] LC_CTYPE=en_US.utf8       LC_NUMERIC=C             
 [3] LC_TIME=en_DK.UTF-8       LC_COLLATE=en_US.utf8    
 [5] LC_MONETARY=en_US.utf8    LC_MESSAGES=en_US.utf8   
 [7] LC_PAPER=en_US.utf8       LC_NAME=C                
 [9] LC_ADDRESS=C              LC_TELEPHONE=C           
[11] LC_MEASUREMENT=en_US.utf8 LC_IDENTIFICATION=C      

time zone: Europe/Berlin
tzcode source: system (glibc)

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] data.table_1.15.99

loaded via a namespace (and not attached):
[1] compiler_4.3.2
```

using `data.table` installed from master as of 2024-07-18 21:36:15 UTC

**Comment by tdhock** [comment]
I translated the example to atime
```r
expr.list <- atime::atime_grid(list(
  AS=paste0("as.data.", c("frame","table"))
), repTables={
  replicate(N, AS(as.list(sapply(letters, function(x) TRUE))), simplify = FALSE)
}, symbol.params="AS")
library(data.table)
atime.result <- atime::atime(
  result=FALSE,
  seconds.limit=0.1, 
  expr.list=expr.list)
plot(atime.result)
```
and I got the result below.
![image](https://github.com/user-attachments/assets/ceed547a-d470-4b8e-a0c7-9b05911bda79)
It suggests that data table uses a constant factor less time, and a constant factor more memory, than data frame.

**Comment by mb706** [comment]
I am mainly confused by the fact that the memory can apparently not be freed by gc(). Has this something to do with how data.table overallocates columns?

**Comment by MichaelChirico** [comment]
> Has this something to do with how data.table overallocates columns?

Yes, that's almost certainly the case. You can try `setalloccol()` to be a smaller number. See also #2950.

**Comment by mb706** [comment]
Investigating this further, I set `datatable.alloccol` to a larger value so I can see its impact on memory usage.

Running the following in `R --vanilla` ends up creating an R session that uses 4GB, which `gc()` refuses to reclaim: 
```r
library("data.table")
options(datatable.alloccol = 1024 * 128)
LENGTH <- 2000
f <- function() { replicate(LENGTH, as.data.table(as.list(sapply(letters, function(x) TRUE))), simplify = FALSE) ; NULL }
f()
replicate(4, gc())->.
```
however, if `LENGTH` is `1970`, `gc()` manages to free all memory; the session temporarily grows to a size of 4GB and then shrinks down again to about 72MB. For numbers between 1970 and 2000 I get conflicting results.

 Is this a problem in data.table, that it sometimes allocates memory in a way that R can not free any more, or is this an R problem?

**Comment by aitap** [comment]
A clue might be that this seems to go away if I lower glibc's [`mmap` threshold](https://www.gnu.org/software/libc/manual/html_node/Memory-Allocation-Tunables.html#index-glibc_002emalloc_002emmap_005fthreshold):
```
$ MALLOC_MMAP_THRESHOLD_=1000 $Rdevel -q -s -e '
library("data.table")
options(datatable.alloccol = 1024 * 128)
LENGTH <- 2000
f <- function() { replicate(LENGTH, as.data.table(as.list(sapply(letters, function(x) TRUE))), simplify = FALSE) ; NULL }
f()
replicate(4, gc())->.
system(paste("ps -q", Sys.getpid(), "-o rss="))
'
NULL
72544
$ $Rdevel -q -s -e '
library("data.table")
options(datatable.alloccol = 1024 * 128)
LENGTH <- 2000
f <- function() { replicate(LENGTH, as.data.table(as.list(sapply(letters, function(x) TRUE))), simplify = FALSE) ; NULL }
f()
replicate(4, gc())->.
system(paste("ps -q", Sys.getpid(), "-o rss="))
'
NULL
4161392
```

Running [`mtrace`](https://www.gnu.org/software/libc/manual/html_node/Tracing-malloc.html) shows a lot of small vectors being allocated by `Rf_allocVector` and being placed in the "small heap" memory area (on amd64 Linux, addresses starting with `0x55...`, as opposed to `0x7f...` for "large" heap allocations backed by `mmap`). Unlike the `mmap`-based allocations that can be returned to the operating system independently from each other, the "small heap" memory area (the data segment) grows and shrinks linearly as a whole using the `sbrk` syscall, and the heap manager may be unable to shrink the data segment if there is a surviving allocation at the far end of it.

The heap fragmentation can be further confirmed by looking at `/proc/self/maps` from inside R:
```
> grepv('heap', readLines('/proc/self/maps'))
[1] "55a72e1c8000-55a80cd26000 rw-p 00000000 00:00 0                          [heap]"
> (0x55a80cd26000 - 0x55a72e1c8000) / 1e9
[1] 3.736461
```
Here, the data segment is responsible for the 3.7 GB taken by the R process.

**Comment by mllg** [comment]
@kalibera can you maybe take a look at the last example w/o `data.table` in https://github.com/Rdatatable/data.table/issues/6301#issuecomment-2295263581? Is this known and expected?

**Comment by kalibera** [comment]
Yes, fragmentation is a plausible explanation for such behavior. As long as the memory management is non-moving, which it is (malloc, R's gc), we can't really avoid it, and one could create artificial tests to generate different kinds of fragmentation. Memory allocators are typically tuned using application benchmarks that should be realistic representations of real applications, but unsurprisingly one may come up with an artificial test that behaves poorly in some sense.

If the default malloc implementation doesn't work well for some important application one needs to run, it is possible either to tune things for that application (as shown above) or use an alternative malloc implementation.  On Linux, one can do this using LD_PRELOAD tricks (e.g. jemalloc, tcmalloc, mimalloc, ...) - some may work better for some applications than the default.

**Comment by aitap** [comment]
There's not much that `data.table` can do about the sub-optimal allocator behaviour.
